consensus/src/experimental/commit_phase.rs

```diff
@@ -0,0 +1,338 @@
+ decision message helps the slower nodes to quickly catch up without
+ having to collect the signatures.
+ */
+
+#[derive(Clone)]
+struct PendingBlocks {
+    vecblocks: Vec<ExecutedBlock>,
+    ledger_info_sig: LedgerInfoWithSignatures,
+}
+
+impl PendingBlocks {
+    pub fn new(vecblocks: Vec<ExecutedBlock>, ledger_info_sig: LedgerInfoWithSignatures) -> Self {
+        Self {
+            vecblocks,
+            ledger_info_sig,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum CommitPhaseMessage {
+    CommitVote(Author, LedgerInfo, Ed25519Signature),
+    CommitDecision(LedgerInfoWithSignatures),
+}
+
+pub struct CommitPhase {
+    commit_channel_recv: Receiver<(Vec<ExecutedBlock>, LedgerInfoWithSignatures)>,
+    execution_proxy: Arc<dyn StateComputer>,
+    local_cache: HashMap<HashValue, LedgerInfoWithSignatures>,
+    local_queue: VecDeque<PendingBlocks>,
+    commit_msg_sender: channel::Sender<CommitPhaseMessage>,
+    commit_msg_receiver: channel::Receiver<ConsensusMsg>,
+    verifier: ValidatorVerifier,
+    safety_rules: Arc<Mutex<MetricsSafetyRules>>,
+    author: Author,
+    committed_round: Round,
+}
+
+pub async fn commit(
+    execution_proxy: &Arc<dyn StateComputer>,
+    vecblock: &[ExecutedBlock],
+    ledger_info: &LedgerInfoWithSignatures,
+) -> Result<(), ExecutionError> {
+    // have to maintain the order.
+    execution_proxy
+        .commit(
+            &vecblock
+                .iter()
+                .map(|eb| Arc::new(eb.clone()))
+                .collect::<Vec<Arc<ExecutedBlock>>>(),
+            ledger_info.clone(),
+        )
+        .await
+}
+
+impl CommitPhase {
+    pub fn new(
+        commit_channel_recv: Receiver<(Vec<ExecutedBlock>, LedgerInfoWithSignatures)>,
+        execution_proxy: Arc<dyn StateComputer>,
+        commit_msg_sender: channel::Sender<CommitPhaseMessage>,
+        commit_msg_receiver: channel::Receiver<ConsensusMsg>,
+        verifier: ValidatorVerifier,
+        safety_rules: Arc<Mutex<MetricsSafetyRules>>,
+        author: Author,
+    ) -> Self {
+        Self {
+            commit_channel_recv,
+            execution_proxy,
+            local_cache: HashMap::<HashValue, LedgerInfoWithSignatures>::new(),
+            local_queue: VecDeque::<PendingBlocks>::new(),
+            commit_msg_sender,
+            commit_msg_receiver,
+            verifier,
+            safety_rules,
+            author,
+            committed_round: 0,
+        }
+    }
+
+    /// Notified when receiving a commit vote message
+    pub async fn process_commit_vote(&mut self, commit_vote: &CommitVote) -> anyhow::Result<()> {
+        let li = commit_vote.ledger_info();
+
+        if li.commit_info().round() < self.committed_round {
+            return Ok(()); // we ignore the message
+        }
+
+        // verify the signature
+        commit_vote.verify(&self.verifier)?;
+
+        let executed_state_hash = li.commit_info().executed_state_id();
```

Comment:
I think we have to use ledger_info().hash(), malicious signer can sign the same executed id but with other fields being different to result in invalid 2f+1 signatures